UnitTests/Diagnostics/Test_Guard.Array.cs

```diff
@@ -0,0 +1,182 @@
+        {
+            Guard.HasSizeOver(new int[1], 4, nameof(Test_Guard_HasSizeAtLeast_ArrayFail));
+        }
+
+        [TestCategory("Guard")]
+        [TestMethod]
+        public void Test_Guard_HasSizeLessThan_ArrayOk()
+        {
+            Guard.HasSizeLessThan(new int[1], 5, nameof(Test_Guard_HasSizeLessThan_ArrayOk));
+        }
+
+        [TestCategory("Guard")]
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        public void Test_Guard_HasSizeLessThan_ArrayEqualFail()
+        {
+            Guard.HasSizeLessThan(new int[4], 4, nameof(Test_Guard_HasSizeLessThan_ArrayEqualFail));
+        }
+
+        [TestCategory("Guard")]
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        public void Test_Guard_HasSizeLessThan_ArrayGreaterFail()
+        {
+            Guard.HasSizeLessThan(new int[6], 4, nameof(Test_Guard_HasSizeLessThan_ArrayGreaterFail));
+        }
+
+        [TestCategory("Guard")]
+        [TestMethod]
+        public void Test_Guard_HasSizeLessThanOrEqualTo_ArrayOk()
+        {
+            Guard.HasSizeLessThanOrEqualTo(new int[1], 5, nameof(Test_Guard_HasSizeLessThanOrEqualTo_ArrayOk));
+            Guard.HasSizeLessThanOrEqualTo(new int[5], 5, nameof(Test_Guard_HasSizeLessThanOrEqualTo_ArrayOk));
+        }
+
+        [TestCategory("Guard")]
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        public void Test_Guard_HasSizeLessThanOrEqualTo_ArrayFail()
+        {
+            Guard.HasSizeLessThanOrEqualTo(new int[8], 4, nameof(Test_Guard_HasSizeLessThanOrEqualTo_ArrayFail));
+        }
+
+        [TestCategory("Guard")]
+        [TestMethod]
+        public void Test_Guard_HasSizeEqualToArray_ArrayOk()
+        {
+            Guard.HasSizeEqualTo(new int[1], new int[1], nameof(Test_Guard_HasSizeEqualToArray_ArrayOk));
+        }
+
+        [TestCategory("Guard")]
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        public void Test_Guard_HasSizeEqualToArray_ArrayFail()
+        {
+            Guard.HasSizeEqualTo(new int[8], new int[2], nameof(Test_Guard_HasSizeEqualToArray_ArrayFail));
+        }
+
+        [TestCategory("Guard")]
+        [TestMethod]
+        public void Test_Guard_HasSizeLessThanOrEqualToArray_ArrayOk()
+        {
+            Guard.HasSizeLessThanOrEqualTo(new int[2], new int[5], nameof(Test_Guard_HasSizeLessThanOrEqualToArray_ArrayOk));
+            Guard.HasSizeLessThanOrEqualTo(new int[4], new int[4], nameof(Test_Guard_HasSizeLessThanOrEqualToArray_ArrayOk));
+        }
+
+        [TestCategory("Guard")]
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        public void Test_Guard_HasSizeLessThanOrEqualToArray_ArrayFail()
+        {
+            Guard.HasSizeLessThanOrEqualTo(new int[8], new int[2], nameof(Test_Guard_HasSizeLessThanOrEqualToArray_ArrayFail));
+        }
```

Comment:
Yup, those are two separate functions:

- `Guard.IsInRangeFor` checks that the input index is valid for the target collection.
- `Guard.HasSizeLessThanOrEqualTo` checks that the size of the source collection is less than or equal to that of the destination collection. Basically, this can be used to check whether you can copy items from one collection to the other.